### PR TITLE
read files from asar archives in custom protocol

### LINKED ISSUE #1069: Can not load files in asar archives in custom protocols
In our app, we reference all of assets with url with different protocols.
I'm currently working on build the distribution app and want to use asar to make a package of the whole app.

We load the start page like this:
```javascript
BrowserWindow.loadUrl('asar:' + __dirname + '/static/dashboard.html');
```
That part alone works, but as soon as I added a protocol.registerProtocol for 'asar' protocol the above page could not be load:
```javascript
    Protocol.registerProtocol('asar', function(request) {
        //do some work with request url
        return new Protocol.RequestFileJob(url);
    });
```

In the console it says: 
```plain
Failed to load resource: asar:/path/to/MyApp.app/Contents/Resources/app.asar/static/dashboard.html
```

The same code would work fine with normal `app` folder even though I use 'asar' protocols for all my asset reference urls.

I suspect that `Protocol.RequestFileJob(url)` does not work for inside asar package, but not sure how to dig deeper.

### PR DESCRIPTION
**Pull request by deepak1556**
fixes #1069

### CODE REVIEW

File: atom/browser/net/adapter_request_job.cc [hunk 1]
```diff
@@ -88,13 +90,32 @@ void AdapterRequestJob::CreateStringJobAndStart(const std::string& mime_type,
...
+    real_job_ = new net::URLRequestFileJob(
+        request(),
+        network_delegate(),
+        path,
+        content::BrowserThread::GetBlockingPool()->
+            GetTaskRunnerWithShutdownBehavior(
+                base::SequencedWorkerPool::SKIP_ON_SHUTDOWN));
+  } else {
+    typedef std::shared_ptr<asar::Archive> asar_ptr;
+    asar_ptr archive = asar::GetOrCreateAsarArchive(asar_path);
```
Comment (zcbenz) [inline atom/browser/net/adapter_request_job.cc]: You can use `auto` here if you don't want to type long type name.

File: atom/browser/net/adapter_request_job.cc [hunk 2]
```diff
@@ -88,13 +90,32 @@ void AdapterRequestJob::CreateStringJobAndStart(const std::string& mime_type,
...
+    if (!archive)
+      real_job_ = new net::URLRequestErrorJob(request(), network_delegate(),
+                                         net::ERR_FILE_NOT_FOUND);
```
Comment (zcbenz) [inline atom/browser/net/adapter_request_job.cc]: The padding should be aligned with the `(`.

**Comment by deepak1556** [comment]
Fixed lint , thanks!

File: atom/browser/net/adapter_request_job.cc [hunk 3]
```diff
@@ -88,13 +90,33 @@ void AdapterRequestJob::CreateStringJobAndStart(const std::string& mime_type,
...
+        content::BrowserThread::GetBlockingPool()->
+            GetTaskRunnerWithShutdownBehavior(
+                base::SequencedWorkerPool::SKIP_ON_SHUTDOWN));
+  } else {
+    auto archive = asar::GetOrCreateAsarArchive(asar_path);
+    if (!archive)
+      real_job_ = new net::URLRequestErrorJob(
+        request(),
+        network_delegate(),
+        net::ERR_FILE_NOT_FOUND);
```
Comment (deepak1556) [inline atom/browser/net/adapter_request_job.cc]: had to change this way to fix cpplint warning of 80 characters :(

**Comment by zcbenz** [comment]
Thanks!

**Pull request merged**